gems/aws-sdk-core/lib/aws-sdk-core/plugins/telemetry.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module Aws
+  module Plugins
+    # @api private
+    class Telemetry < Seahorse::Client::Plugin
+      option(
+        :telemetry_provider,
+        default: Aws::Telemetry::NoOpTelemetryProvider,
+        doc_type: Aws::Telemetry::TelemetryProviderBase,
+        rbs_type: Aws::Telemetry::TelemetryProviderBase,
+        docstring: <<~DOCS) do |_cfg|
+Allows you to provide a telemetry provider, which is used
+to emit telemetry data. By default, uses `NoOpTelemetryProvider`
+which will not record or emit any telemetry data. The SDK currently supports
+OpenTelemetry (OTel) as a telemetry provider. To use the OTel provider, 
+require the`opentelemetry-sdk` gem and then, pass in an instance of a 
+`Aws::Telemetry::OTelProvider` for telemetry provider.
+        DOCS
+        Aws::Telemetry::NoOpTelemetryProvider.new
+      end
+
+      def after_initialize(client)
+        validate_telemetry_provider(client.config)
+      end
+
+      def validate_telemetry_provider(config)
+        unless config.telemetry_provider.is_a?(Aws::Telemetry::TelemetryProviderBase)
+          raise ArgumentError,
+                'Must provide a telemetry provider for the '\
+                '`telemetry_provider` configuration option.'
+        end
+      end
+
+      class Handler < Seahorse::Client::Handler
+        def call(context)
+          service_id = service_id(context)
+          attributes = {
+            'rpc.system' => 'aws-api',
+            'rpc.service' => service_id,
+            'rpc.method' => context.operation.name,
+            'code.function' => context.operation_name.to_s,
+            'code.namespace' => 'Aws::Plugins::Telemetry'
+          }
+          context.tracer.in_span(
+            parent_span_name(context, service_id),
+            attributes: attributes,
+            kind: Aws::Telemetry::SpanKind::CLIENT
+          ) do
+            @handler.call(context)
+          end
+        end
+
+        private
+
+        def service_id(context)
+          context.config.api.metadata['serviceId'] ||
```

Comment:
I think serviceId exists on all models now, you should check.

Reply:
Importexport is deprecated and uses sigv2. We could add a customization to add service id to it just to make this work. All active services should have service id. I think we could actually delete this from the repo.

Reply:
we have similar code that falls back to service abbreviation elsewhere right? I think its fine to just keep this or rather than add a customization (just a slight preference not a strong feeling).  Its true that all services going forward must have serviceId, so maybe it does make sense to just customize + remvove these special cases.
Or maybe we can put this fallback logic once, in our build tools somewhere so that it ensures the api in ruby runtime always has serviceId?

Reply:
All services should have serviceId backfilled. Other than importexport, which ones don't? They should arguably be removed.

Reply:
The logic we have I believe in release tools and I think maybe in gem name in build tools. I would be in favor of adding to customizations.rb to add service id for the few services missing it and then get them removed.

Reply:
Per our discussion, we are keeping the fall back here to cover cases with old services, new core scenario. That being said, added customization in the build_tools to resolve the missing serviceId for those three services.